### PR TITLE
Make EventReader a SystemParam

### PR DESCRIPTION
**Pull request by TheRawMeatball**
This commit makes `EventReader` a `SystemParam` to significantly cut down on boilerplate, and creates a simplified `ManualEventReader` to allow for use cases with direct access to Resources. As a necessary side effect, this commit also adds generic type support to `#[derive(SystemParam)]`. Before a merge, I'd also suggest cutting the `EventReader` API down to only `iter()` and `iter_with_id()`, since rust iterators provide more idiomatic solutions to the rest of the functions.

### CODE REVIEW

**Comment by cart** [comment]
Very cool. For _most_ things I don't want to encourage special-cased SystemParams as they increase api surface, but events are "core" enough that I think its worth investing in ergonomics.

> As a necessary side effect, this commit also adds generic type support to #[derive(SystemParam)]

Brilliant. Very glad to have that back. I punted it when adding lifetimes to SystemParams because I had no clue how to filter the generics. I'll add the pattern you used here to my repertoire  :smile: 

> Before a merge, I'd also suggest cutting the EventReader API down to only iter() and iter_with_id(), since rust iterators provide more idiomatic solutions to the rest of the functions.

I'm sold on that!

**Comment by TheRawMeatball** [comment]
Before merging, I'd suggest changing or extending the documentation. I'd do it, but I don't know what to add, I just have a feeling the docs emphasize the `Events` too much. Maybe not a large problem since the examples are changed, but still

**Comment by alice-i-cecile** [comment]
I've read over the changes, and like this PR a lot. The reduced API surface in particular is a good direction: I like being able to rely on core functionality more.

1.  I discussed the "consuming event" pattern in #1157: it looks like we should be able to do that with `raw_events.iter().next()` pretty elegantly, so this PR should close that issue.

2. IMO, a second event reader should be added to the docstring example, to clearly demonstrate that multiple event readers can read the same event.

3. We should have an example for how to manually handle your own events, possibly using the consuming event pattern. It's not obvious how to do so from the current examples / docs and you need to dig into how `add_event` works to figure it out.

**Comment by alice-i-cecile** [comment]
I would also like to see `Events` be able to be called in system parameters directly. Rather than ` mut my_events: ResMut<Events<MyEvent>>`, use `mut my_events: Events<MyEvent>`. This allows us to expose a parallel-safe API for events (like our current Commands implementation) that only allows users to write to the end, which lets us run multiple systems that generate the same event in parallel.

In the rare cases where you want full read or write access (like when processing raw events), you can use `Res<MyEvent>` instead.

**Comment by tigregalis** [comment]
I love this change. The boilerplate was difficult to write from scratch from memory without looking at a reference implementation each time. This feels really intuitive (and cuts down on a lot of boilerplate).

**Comment by TheRawMeatball** [comment]
> I would also like to see Events be able to be called in system parameters directly.

Hmm, `Events` is the backing resource, so I feel like using it directly would be a bad idea. Perhaps a `EventWriter`? If that is OK, I can implement it on [this](https://github.com/TheRawMeatball/bevy/tree/clean-buffer-rebase) branch, but it'll have to wait until Ratysz s new scheduler lands. So, I'd recommend merging the reader first, and after the new scheduler is merged I'll open a PR for a writer as well

**Comment by tigregalis** [comment]
> 
> 
> > I would also like to see Events be able to be called in system parameters directly.
> 
> Hmm, `Events` is the backing resource, so I feel like using it directly would be a bad idea. Perhaps a `EventWriter`? If that is OK, I can implement it on [this](https://github.com/TheRawMeatball/bevy/tree/clean-buffer-rebase) branch, but it'll have to wait until Ratysz s new scheduler lands. So, I'd recommend merging the reader first, and after the new scheduler is merged I'll open a PR for a writer as well

`EventReader` + `EventWriter` makes a lot of sense.

**Comment by TheRawMeatball** [comment]
The `EventWriter` is now implemented at https://github.com/TheRawMeatball/bevy/tree/event-writer! With possibly more things taking the same route `Commands` does, I might also simplify `&mut Commands` etc to just be `Commands`

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -209,7 +209,11 @@ pub fn winit_runner_with(mut app: App, mut event_loop: EventLoop<()>) {
         *control_flow = ControlFlow::Poll;
 
         if let Some(app_exit_events) = app.resources.get_mut::<Events<AppExit>>() {
-            if app_exit_event_reader.latest(&app_exit_events).is_some() {
+            if app_exit_event_reader
+                .iter(&app_exit_events)
+                .next()
```
Comment (cart) [inline crates/bevy_winit/src/lib.rs:214]: latest returned the "last" item in the iterator, so this should be `last()`

File: crates/bevy_ui/src/flex/mod.rs [hunk 2]
```diff
@@ -219,7 +218,7 @@ pub fn flex_node_system(
         1.
     };
 
-    if scale_factor_reader.latest(&scale_factor_events).is_some() {
+    if scale_factor_events.iter().next().is_some() {
```
Comment (cart) [inline crates/bevy_ui/src/flex/mod.rs:221]: latest returned the "last" item in the iterator, so this should be `last()`

**Comment by TheRawMeatball** [comment]
I changed it back since it was a minor change, though I don't think its necessary since we're only checking if an event occured, so its probably fine to use either. I originally chose next since it made simpler code.

**Comment by cart** [comment]
Changing next() to latest() (or next_back()) isn't a frivolous no-op.  Calling next() in the context above results in:
1) returning the oldest event
2) updating the index to ignore all newer events in the buffers

This would result in dropping newer events if we queue up multiple in the same frame. If someone called `events.send(A)` and then `events.send(B)`, in this case B should clearly be the final state we end up in. Would you not consider ending in state A a bug?

**Comment by TheRawMeatball** [comment]
Hmm, I thought it wouldn't matter since all events are erased at the end of the frame, do they last longer? Also, since we aren't looking at the state in the events, wouldn't A and B be identical

**Comment by cart** [comment]
Bawhaha you are absolutely right. I totally missed that we are just calling `is_some()` (and reading the "latest" value from the window directly).

Thanks for keeping me honest :smile:

**Comment by cart** [comment]
(it really is a frivolous no-op)

**Pull request merged**